src/server.ts

```diff
@@ -57,9 +57,21 @@ export function createServerWithTools(options: Options): Server {
       };
     }
 
+    const modalStates = context.modalStates().map(state => state.type);
+    if ((tool.clearsModalState && !modalStates.includes(tool.clearsModalState)) ||
+        (!tool.clearsModalState && modalStates.length)) {
+      const text = [
+        `Tool "${request.params.name}" does not handle the modal state.`,
+        ...context.modalStatesMarkdown(),
+      ].join('\n');
+      return {
+        content: [{ type: 'text', text }],
+        isError: true,
```

Comment:
I'm a little wary that the model perceives this as a broken tool, not as wrong usage, and that it stops using the tool entirely. https://modelcontextprotocol.io/docs/concepts/tools#error-handling-2 mentions something about human intervention, which could be interpreted like that. Let's keep an eye out.